cmd/clusterawsadm/iamservice/service.go

```diff
@@ -0,0 +1,500 @@
+	data, err := json.Marshal(rawdata)
+	if err != nil {
+		return errors.Wrapf(err, "corrupt policy document format for policy \"%s\"", res.ManagedPolicyName)
+	}
+	create, err := client.CreatePolicy(&iam.CreatePolicyInput{
+		Description:    &res.Description,
+		PolicyDocument: aws.String(string(data)),
+		PolicyName:     &res.ManagedPolicyName,
+		Tags:           tgs,
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeEntityAlreadyExistsException:
+				klog.Warningf("policy \"%s\" already exists", res.ManagedPolicyName)
+				policies, err := listpolicies(client)
+				if err != nil {
+					return err
+				}
+				for _, policy := range policies {
+					if *policy.PolicyName == res.ManagedPolicyName {
+						return attachRolesToPolicy(policy, res.Roles, client)
+					}
+				}
+			default:
+				return errors.Wrapf(err, "failed to create CAPA managed IAM policy \"%s\"", res.ManagedPolicyName)
+			}
+		}
+	}
+	err = attachRolesToPolicy(create.Policy, res.Roles, client)
+	if err != nil {
+		return err
+	}
+	klog.V(2).Infof("created \"%s\" CAPA managed IAM policy", res.ManagedPolicyName)
+	return nil
+}
+
+func attachRoleToInstanceProf(resource go_cfn.Resource, client *iam.IAM) error {
+	res := resource.(*cfn_iam.InstanceProfile)
+	roleName, err := getRoleName(res.Roles[0])
+	if err != nil {
+		return err
+	}
+	_, err = client.AddRoleToInstanceProfile(&iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: &res.InstanceProfileName,
+		RoleName:            &roleName,
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeEntityAlreadyExistsException, iam.ErrCodeLimitExceededException:
+				klog.Warningf("instance profile \"%s\" is already attached to its IAM role", res.InstanceProfileName)
+			default:
+				return errors.Wrapf(err, "failed to attach instance profile \"%s\" to IAM role \"%s\"", res.InstanceProfileName, roleName)
+			}
+		}
+	}
+	return nil
+}
+
+func attachPoliciesToRole(rolename *string, awsManagedPolicies []string, client *iam.IAM) error {
+	if awsManagedPolicies == nil {
+		// klog.Warningf("no policies defined to attach to the IAM role \"%s\"", *rolename) // TODO
+		return nil
+	}
+	for _, policyArn := range awsManagedPolicies {
+		// making a copy of policyArn to avoid implicit memory aliasing
+		policy := policyArn
+		_, err := client.AttachRolePolicy(&iam.AttachRolePolicyInput{
+			RoleName:  rolename,
+			PolicyArn: &policy,
+		})
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case iam.ErrCodeEntityAlreadyExistsException:
+					klog.Warningf("IAM role \"%s\" is already attached to policy", *rolename) // TODO should we output the policy arn? how safe is it
+					continue
+				default:
+					return errors.Wrapf(err, "failed to attach IAM role \"%s\" to policy", *rolename) // TODO should we output the policy arn? how safe is it
```

Comment:
Thinking about where this command should be running, I can think of a couple environments:

First, a user's machine, while they set up a management cluster. I think outputting an ARN here is fine. They have some sort credentials on their machine, which are the higher risk asset.


Next, CI, which might be spinning up resources, either for testing or as part of a larger provisioning system. In this case, the ARNs would probably get output into logs. Hopefully, the CI is using short-lived credentials, but again, those would be the more valuable asset.

It's true that policy ARNs could be observed and an attacker could attempt to edit the policy to expand permissions. However, I'm not sure the actual ARN is all that important, since what really matters are the permissions to actually overwrite the policy the ARN points to.

Does that make sense?

Reply:
Yeah that make sense :) I'll make the change to include the ARNs in the outputs above.